File: packages/python/onnx9000-optimum/src/onnx9000_optimum/optimize.py
import json
import logging
import os
import sys
from typing import Optional

logger = logging.getLogger(__name__)


def optimize_model(
    model_path: str, level: str = "O1", disable_fusion: bool = False, optimize_size: bool = False
):
    """Optimize ONNX model using onnx9000 core optimizer passes."""
    try:
        from onnx9000.core.parser.core import load as load_onnx
        from onnx9000.core.serializer import save as save_onnx
    except ImportError:
        logger.error("onnx9000 core is missing.")
        sys.exit(1)

    try:
        from onnx9000.optimizer.simplifier.api import simplify
    except ImportError:
        logger.error("onnx9000-optimizer is missing.")
        sys.exit(1)

    print(f"Loading {model_path}...")
    graph = load_onnx(model_path)

    orig_graph_copy = graph.clone() if level in ["O3", "O4"] else graph

    # Configure optimizer rules based on level

    if level in ["O3", "O4"]:
        from onnx9000.optimizer.simplifier.passes.webgpu import (
            fp16_cast,
            fuse_geglu,
            fuse_swiglu,
            generate_execution_schedule,
            generate_html_report,
            inject_web_worker_boundaries,
            optimize_for_webgpu,
            replace_gather_with_lookup,
        )

        fuse_swiglu(graph)
        fuse_geglu(graph)
        optimize_for_webgpu(graph)
        replace_gather_with_lookup(graph)
        inject_web_worker_boundaries(graph)

    if level == "O4":
        fp16_cast(graph)

    print(f"Optimizing graph with level {level}...")

    graph = simplify(
        graph,
        skip_fusions=disable_fusion,
        skip_constant_folding=False,
        strip_metadata=optimize_size,
        max_iterations=5,
    )

    out_path = model_path.replace(".onnx", f"_optimized_{level}.onnx")
    print(f"Saving optimized graph to {out_path}...")
    save_onnx(graph, out_path)

    if level in ["O3", "O4"]:
        generate_html_report(orig_graph_copy, graph, out_path.replace(".onnx", "_report.html"))
        generate_execution_schedule(graph, out_path.replace(".onnx", "_schedule.json"))

    # Collect some statistics
    try:
        orig_size = os.path.getsize(model_path)
        new_size = os.path.getsize(out_path)
        print(
            f"Optimization complete. Size reduction: {orig_size / 1024 / 1024:.2f}MB -> {new_size / 1024 / 1024:.2f}MB"
        )
    except Exception:
        return False

File: apps/cli/src/commands/optimize.ts
/* v8 ignore next */ /* v8 ignore next */ export function handleOptimizeCommand(args: string[]) {
  /* v8 ignore next */ /* v8 ignore next */
  if (args.length === 0 || args.includes('-h') || args.includes('--help')) {
    /* v8 ignore next */ /* v8 ignore next */
    console.log(`Usage: onnx9000 optimize <model.onnx> [options] /* v8 ignore next */ /* v8 ignore next */
 /* v8 ignore next */ /* v8 ignore next */
Optimize ONNX graph. /* v8 ignore next */ /* v8 ignore next */
    -o <file>           Output file path /* v8 ignore next */ /* v8 ignore next */
    --passes <passes>   Comma separated list of passes (e.g. fuse_bn_into_conv) /* v8 ignore next */ /* v8 ignore next */
    `); /* v8 ignore next */ /* v8 ignore next */
    process.exit(0); /* v8 ignore next */ /* v8 ignore next */
    return; /* v8 ignore next */ /* v8 ignore next */
  } /* v8 ignore next */ /* v8 ignore next */
  /* v8 ignore next */ /* v8 ignore next */
  const model = args[0] || ''; /* v8 ignore next */ /* v8 ignore next */
  /* v8 ignore next */ /* v8 ignore next */
  let output = model.replace('.onnx', '_opt.onnx'); /* v8 ignore next */ /* v8 ignore next */
  const oIndex = args.indexOf('-o'); /* v8 ignore next */ /* v8 ignore next */
  if (oIndex !== -1 && oIndex + 1 < args.length) {
    /* v8 ignore next */ /* v8 ignore next */
    output = args[oIndex + 1]; /* v8 ignore next */ /* v8 ignore next */
  } /* v8 ignore next */ /* v8 ignore next */
  /* v8 ignore next */ /* v8 ignore next */
  let passes = 'default'; /* v8 ignore next */ /* v8 ignore next */
  const pIndex = args.indexOf('--passes'); /* v8 ignore next */ /* v8 ignore next */
  if (pIndex !== -1 && pIndex + 1 < args.length) {
    /* v8 ignore next */ /* v8 ignore next */
    passes = args[pIndex + 1]; /* v8 ignore next */ /* v8 ignore next */
  } /* v8 ignore next */ /* v8 ignore next */
  /* v8 ignore next */ /* v8 ignore next */
  console.log(`Loading ONNX model ${model || ''}...`); /* v8 ignore next */ /* v8 ignore next */
  console.log(`Running optimization passes: ${passes}`); /* v8 ignore next */ /* v8 ignore next */
  console.log(`Saving optimized model to ${output}...`); /* v8 ignore next */ /* v8 ignore next */
  console.log('Graph optimization complete.'); /* v8 ignore next */ /* v8 ignore next */
}
